### PR TITLE
MINOR: Stabilize ApplicationResetIntegrationTest

### PR DESCRIPTION
**Pull request by vvcephei**
Increase retry backoff time and make Streams startup synchronous to improve test stability.

### CODE REVIEW

File: src/main/java/io/confluent/examples/streams/ApplicationResetExample.java [hunk 1]
```diff
@@ -143,13 +144,13 @@ public static void main(final String[] args) {
...
     final boolean doReset = args.length > 1 && args[1].equals("--reset");
-    final KafkaStreams streams = run(doReset, streamsConfiguration);
+    final KafkaStreams streams = run(doReset, streamsConfiguration, false);
 
     // Add shutdown hook to respond to SIGTERM and gracefully close Kafka Streams
     Runtime.getRuntime().addShutdownHook(new Thread(streams::close));
   }
 
-  public static KafkaStreams run(final boolean doReset, final Properties streamsConfiguration) {
+  public static KafkaStreams run(final boolean doReset, final Properties streamsConfiguration, final boolean wait) {
```
Comment (mjsax) [inline src/main/java/io/confluent/examples/streams/ApplicationResetExample.java]: nit: `wait` on what? maybe update variable name?

**Comment by vvcephei** [comment]
Thanks for the nit, @mjsax . I've refactored the code to eliminate both boolean flags, actually, and decompose the operation into its granular parts.

WDYT?

File: src/main/java/io/confluent/examples/streams/ApplicationResetExample.java [hunk 2]
```diff
@@ -143,32 +144,47 @@ public static void main(final String[] args) {
...
+  public static KafkaStreams buildKafkaStreams(final Properties streamsConfiguration) {
     // Define the processing topology
     final StreamsBuilder builder = new StreamsBuilder();
     final KStream<String, String> input = builder.stream("my-input-topic");
     input.selectKey((key, value) -> value.split(" ")[0])
-      .groupByKey()
-      .count()
-      .toStream()
-      .to("my-output-topic", Produced.with(Serdes.String(), Serdes.Long()));
+         .groupByKey()
```
Comment (mjsax) [inline src/main/java/io/confluent/examples/streams/ApplicationResetExample.java]: nit: we use 2 space indention in CP repos (4 space is for AK)

File: src/test/java/io/confluent/examples/streams/ApplicationResetIntegrationTest.java [hunk 3]
```diff
@@ -133,15 +138,19 @@ public void shouldReprocess() throws Exception {
     //
     // Step 5: Rerun application
     //
-    streams = ApplicationResetExample.run(true, streamsConfiguration);
+    final KafkaStreams streams1 = ApplicationResetExample.buildKafkaStreams(streamsConfiguration);
```
Comment (mjsax) [inline src/test/java/io/confluent/examples/streams/ApplicationResetIntegrationTest.java]: `streams1` is a very bad name -- also, logically, we restart the same application, thus reusing existing variable seems sound to me.

File: src/main/java/io/confluent/examples/streams/ApplicationResetExample.java [hunk 4]
```diff
@@ -143,32 +144,47 @@ public static void main(final String[] args) {
...
 
-    final KafkaStreams streams = new KafkaStreams(builder.build(), streamsConfiguration);
+    return new KafkaStreams(builder.build(), streamsConfiguration);
+  }
 
-    // Delete the application's local state on reset
-    if (doReset) {
-      streams.cleanUp();
-    }
+  public static void startKafkaStreams(final KafkaStreams streams) {
```
Comment (mjsax) [inline src/main/java/io/confluent/examples/streams/ApplicationResetExample.java]: nit: `startKafkaStreamsSync` (or similar -- otherwise unclear what the difference to calling `#start()`

File: src/test/java/io/confluent/examples/streams/ApplicationResetIntegrationTest.java [hunk 5]
```diff
@@ -77,7 +78,11 @@ public void shouldReprocess() throws Exception {
     // Use a temporary directory for storing state, which will be automatically removed after the test.
     streamsConfiguration.put(StreamsConfig.STATE_DIR_CONFIG, TestUtils.tempDirectory().getAbsolutePath());
 
-    KafkaStreams streams = ApplicationResetExample.run(false, streamsConfiguration);
+    streamsConfiguration.put(AdminClientConfig.RETRY_BACKOFF_MS_CONFIG, 1000);
```
Comment (mjsax) [inline src/test/java/io/confluent/examples/streams/ApplicationResetIntegrationTest.java]: Why this change? Seems to be unrelated?

File: src/main/java/io/confluent/examples/streams/ApplicationResetExample.java (hunk 2 above)
Comment (vvcephei) [reply src/main/java/io/confluent/examples/streams/ApplicationResetExample.java]: Ah, sorry about that.

File: src/test/java/io/confluent/examples/streams/ApplicationResetIntegrationTest.java (hunk 3 above)
Comment (vvcephei) [reply src/test/java/io/confluent/examples/streams/ApplicationResetIntegrationTest.java]: Ok.

File: src/main/java/io/confluent/examples/streams/ApplicationResetExample.java (hunk 4 above)
Comment (vvcephei) [reply src/main/java/io/confluent/examples/streams/ApplicationResetExample.java]: Ok.

File: src/test/java/io/confluent/examples/streams/ApplicationResetIntegrationTest.java (hunk 5 above)
Comment (vvcephei) [reply src/test/java/io/confluent/examples/streams/ApplicationResetIntegrationTest.java]: The other changes are to improve stability in general. This one is specifically addressing a recent test failure in which there was a race between the broker actually completing the cleanup from the reset tool and streams trying to (re)create the internal topics.

Streams only retries 5 times with a backoff time of 100ms, so it only waits 0.5 seconds for the broker to delete the topic, which works some times but not others. Now, it'll wait up to 5 seconds, which should be plenty of time. Of course, if the broker cleans it up sooner, we'll be able to continue the test, in increments of 1 second.

File: src/main/java/io/confluent/examples/streams/ApplicationResetExample.java (hunk 2 above)
Comment (vvcephei) [reply src/main/java/io/confluent/examples/streams/ApplicationResetExample.java]: Oh, actually, this is just aligning the chained calls. Do you think we should prefer to strictly indent by 2 spaces in preference to keeping chained calls aligned?

File: src/test/java/io/confluent/examples/streams/ApplicationResetIntegrationTest.java (hunk 3 above)
Comment (vvcephei) [reply src/test/java/io/confluent/examples/streams/ApplicationResetIntegrationTest.java]: I've refactored it using code blocks so that we can use the same name without re-assigning the same variable.

Hopefully, this is the best of both worlds.

File: src/main/java/io/confluent/examples/streams/ApplicationResetExample.java (hunk 2 above)
Comment (mjsax) [reply src/main/java/io/confluent/examples/streams/ApplicationResetExample.java]: Don't know. I personally would move `.selectKey()` to it's own line and indent with 2 spaces.

File: src/test/java/io/confluent/examples/streams/ApplicationResetIntegrationTest.java [hunk 6]
```diff
@@ -77,33 +81,39 @@ public void shouldReprocess() throws Exception {
...
+      producerConfig.put(ProducerConfig.RETRIES_CONFIG, 0);
+      producerConfig.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, ByteArraySerializer.class);
+      producerConfig.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
+      IntegrationTestUtils.produceValuesSynchronously(inputTopic, inputValues, producerConfig);
+
+      //
+      // Step 3: Verify the application's output data.
+      //
+
+
```
Comment (mjsax) [inline src/test/java/io/confluent/examples/streams/ApplicationResetIntegrationTest.java]: double empty line intended?

**Comment by mjsax** [comment]
Still LGTM. Feel free to merge from my side.

File: src/test/java/io/confluent/examples/streams/ApplicationResetIntegrationTest.java [hunk 7]
```diff
@@ -18,13 +18,10 @@
...
 import kafka.tools.StreamsResetter;
+import org.apache.kafka.clients.admin.AdminClientConfig;
 import org.apache.kafka.clients.consumer.ConsumerConfig;
 import org.apache.kafka.clients.producer.ProducerConfig;
-import org.apache.kafka.common.serialization.ByteArraySerializer;
-import org.apache.kafka.common.serialization.LongDeserializer;
-import org.apache.kafka.common.serialization.Serdes;
-import org.apache.kafka.common.serialization.StringDeserializer;
-import org.apache.kafka.common.serialization.StringSerializer;
+import org.apache.kafka.common.serialization.*;
```
Comment (ableegoldman) [inline src/test/java/io/confluent/examples/streams/ApplicationResetIntegrationTest.java]: nit: avoid wildcard import (IntelliJ strikes again)

File: src/test/java/io/confluent/examples/streams/ApplicationResetIntegrationTest.java (hunk 6 above)
Comment (vvcephei) [reply src/test/java/io/confluent/examples/streams/ApplicationResetIntegrationTest.java]: nope! I'll fix it.

File: src/test/java/io/confluent/examples/streams/ApplicationResetIntegrationTest.java (hunk 7 above)
Comment (vvcephei) [reply src/test/java/io/confluent/examples/streams/ApplicationResetIntegrationTest.java]: Darn it. Thanks for the catch.

**Comment by vvcephei** [comment]
Resolved comments. Waiting for tests to pass before merging.
